pyccl/base.py

```diff
@@ -298,3 +299,332 @@ def increment(self):
 
     def reset(self):
         self.counter = 0
+
+
+def auto_assign(func, sig):
+    """Decorator to automatically assign all parameters as instance attributes.
+    This ought to be applied on constructor methods.
+
+    Arguments:
+        func (``function``):
+            Function which takes the instance as its first argument.
+            All function arguments will be assigned as attributes of the
+            instance.
+        sig (``inspect.Signature``):
+            The signature of the constructor this decorator is applied to.
+            This is needed to distinguish between parent class methods.
+    """
+    @functools.wraps(func)
+    def wrapper(self, *args, **kwargs):
+        bound = sig.bind_partial(self, *args, **kwargs)
+        bound.apply_defaults()
+        bound.arguments.pop("self")
+        for name, value in bound.arguments.items():
+            setattr(self, name, value)
+        # func may now override the attributes we just set
+        func(self, *args, **kwargs)
+    return wrapper
+
+
+class ObjectLock:
+    """Control the lock state (immutability) of a ``CCLObject``."""
+    _locked: bool = False
+    _lock_id: int = None
+
+    def __repr__(self):
+        return f"{self.__class__.__name__}(locked={self.locked})"
+
+    @property
+    def locked(self):
+        """Check if the object is locked."""
+        return self._locked
+
+    @property
+    def active(self):
+        """Check if an unlocking context manager is active."""
+        return self._lock_id is not None
+
+    def lock(self):
+        """Lock the object."""
+        self._locked = True
+        self._lock_id = None
+
+    def unlock(self, manager_id=None):
+        """Unlock the object."""
+        self._locked = False
+        if manager_id is not None:
+            self._lock_id = manager_id
+
+
+class UnlockInstance:
+    """Context manager that temporarily unlocks an immutable instance
+    of ``CCLObject``.
+
+    Parameters:
+        instance (``CCLObject``):
+            Instance of ``CCLObject`` to unlock within the scope
+            of the context manager.
+        mutate (``bool``):
+            If the enclosed function mutates the object, the stored
+            representation is automatically deleted.
+    """
+
+    def __init__(self, instance, mutate=True):
+        self.instance = instance
+        self.mutate = mutate
+        # Define these attributes for easy access.
+        self.id = id(self)
+        self.thread_lock = RLock()
+        # We want to catch and exit if the instance is not a CCLObject.
+        # Hopefully this will be caught downstream.
```

Comment:
What?

Reply:
Foul language maybe, I meant if the object that we try to unlock comes from outside of the CCL library, the devs should catch it downstream of course.